### ISSUE TITLE
Question: Do createCallbacks only accept a single Node as a return value?

### ISSUE THREAD

**Issue by ryansolid** [Freak613/stage0#12]
I was very interested by this approach of having a single data reconciliation method to handle lists. Most libraries tend to iterate over the data maybe up to 3 times to get everything in the right place. But I realized looking at the code since it relies on stepping through the DOM after the fact using next/previous sibling each child must correspond to to a single child. Even if you cheated a bit and returned a document fragment that would append multiple children with a single append action the reconciler has no way of knowing how to progress.

Just confirming that this is the case and wondering if you had any thoughts on how to approach it.

**Comment by leeoniya** [comment]
> and we don't have a tool to control this structure as a single item later(or I don't know it?)

no :(

**Comment by Freak613** [comment]
I just thought that maybe ShadowDOM could be worth to try, as it controls its content as a single item. I'll try to experiment with it later.

**Comment by Freak613** [comment]
But maybe not. Since it looks like element could have only one shadow dom in it. But anyway should try to test it.

I played with Range before, but concluded that it's only for selected text. Maybe I'm wrong and I just didn't figure out how to use it yet.

https://developer.mozilla.org/en-US/docs/Web/API/Range

**Comment by ryansolid** [comment]
Yeah I figured as much I was just curious if you'd considered stuff here since you'd done a really good job of generalizing this even with the not the only child scenario. Could mark the DOM nodes with some made up property almost like a key and iterate (via nextSibling etc) over elements until key changes but I'm not sure if the writing or lookup would be more expensive than just holding the DOM nodes.

ShadowRoot is just a Document Fragment essentually and yeah never had much luck with Ranges. I suspect the performance wouldn't be good enough. Anyway gives me some ideas to try out. Thanks.

**Comment by ryansolid** [comment]
Yeah so I ended up doing what I suggested with keyed nodes with Document Fragments. Seems to work fast enough. It's still a bit of a mess (I also injected in disposal hooks necessary for fine grained libraries like S.js, MobX), but under the scenarios I was working on with KVO libraries I didn't get to do that much down to the metal comparisons. I've incorporated that into the latest version of my Babel Plugin I use for my libraries. When I get around to optimizing and cleaning it up maybe I will submit a pull request here.
